File: dasol/양궁대회.py
def solution(n, info):
    # 냅색 문제 + 백트래킹 문제. 냅색 푼지가 너무 오래되서 기억이 전혀 안났음;;
    apeach_score = sum(10-i for i in range(11) if info[i] != 0)
    
    # 어피치가 확보한 점수를 뺏으면 score의 2배를 얻는 것으로 보면 되고,
    # 어피치가 확보하지 못한 점수를 얻으면 score를 얻는 것으로 보면 됩니다.
    # weight는 각 점수 별로 (어피치가 쏜 화살 수+1)로 정의됩니다.
    weights = [-1] + [x+1 for x in info]
    values = [-1] + [2*(10-i) if info[i] != 0 else (10-i) for i in range(11)]
    
    # 냅색 풀이코드
    A = [[0] * (n + 1) for _ in range(12)]
    for i in range(1, 12):
        for w in range(1, n+1):
            if weights[i] > w:
                A[i][w] = A[i-1][w]
            else:
                A[i][w] = max(A[i-1][w-weights[i]] + values[i], A[i-1][w])

    # 오른쪽 끝 점수를 봅니다.
    # 어피치가 획득했던 점수보다 작거나 같으면 비기거나 진다는 뜻입니다.
    final_score = A[-1][-1] - apeach_score
    if final_score <= 0:
        return [-1]
    
    # 백트래킹 해서 각 점수에 화살을 얼마나 쏴야 할지 봅니다.
    answer = [0] * 11
    w = n
    for i in range(11, 0, -1):
        # A[i][w] 값이 A[i-1][w-weights[i]] + values[i]와 같으면
        # i점에 "적어도" weights[i]개의 화살을 쐈다는 것입니다.
        if A[i][w] == A[i-1][w-weights[i]] + values[i]:
            answer[i-1] = weights[i]
            w = w-weights[i]
    
    answer[-1] += (n - sum(answer))
    return answer

File: younghyo/양궁대회.java

// 프로그래머스 양궁대회
// 모르겠어요
// 백트래킹 완전탐색

public class YangGoong {
	
	boolean[] visited = new boolean[11]; // 점수 확인

	int total = 0; // 점수 합
	int max = 0; // 점수 차

	int[] need = new int[11];
	int[] vic = new int[11];

	public int[] solution(int n, int[] info) {
		int[] answer = {};
		for (int i = 0; i < 11; i++) {
			need[i] = info[i] + 1;
			if (info[i] != 0) {
				total += 10 - i;
			}
		}
		need[10] = 0;
		rec(0, 11, n);
		if (max == 0) {
			return new int[] { -1 };
		}
		return vic;
	}

	public void rec(int start, int r, int n) {
		if (r == 0) {
			int sum = 0;
			int[] arr = new int[11];
			for (int i = 0; i < 11; i++) {
				if (visited[i]) {
					arr[i] = need[i];
					if (arr[i] == 1) {
						sum += 10 - i;
					} else {
						sum += (10 - i) * 2;
					}
				}
			}
			arr[10] = n;
			if (max == sum - total) {
				for (int i = 10; i >= 0; i--) {
					if (vic[i] > arr[i]) {
						break;
					} else if (vic[i] == arr[i]) {
						continue;
					} else {
						vic = arr;
					}
				}
			} else if (max < sum - total) {
				max = Math.max(sum - total, max);
				vic = arr;
			}
			return;
		}
		if (n == 0) {
			int sum = 0;
			int[] arr = new int[11];
			for (int i = 0; i < 11; i++) {
				if (visited[i]) {
					arr[i] = need[i];
					if (arr[i] == 1) {
						sum += 10 - i;
					} else {
						sum += (10 - i) * 2;
					}
				}
			}
			if (max == sum - total) {
				for (int i = 10; i >= 0; i--) {
					if (vic[i] > arr[i]) {
						break;
					} else if (vic[i] == arr[i]) {
						continue;
					} else {
						vic = arr;
					}
				}
			} else if (max < sum - total) {
				max = Math.max(sum - total, max);
				vic = arr;
			}
			return;
		}

		for (int i = start; i < 11; i++) {
			if (n >= need[i]) {
				visited[i] = true;
				n -= need[i];
				rec(start + 1, r - 1, n);
				visited[i] = false;
				n += need[i];
			}

		}
	}
}

// 
